compiler/src/iree/compiler/Codegen/Common/LinkTuningSpecsPass.cpp

```diff
@@ -61,6 +62,210 @@ static bool consumesInputOp(NamedSequenceOp op) {
   return false;
 }
 
+static bool hasConsumedArgument(ForeachMatchOp op) {
+  Value operand = op->getOperand(0);
+  if (auto blockArg = dyn_cast<BlockArgument>(operand)) {
+    Operation *parentOp = blockArg.getOwner()->getParentOp();
+    if (auto namedSequenceOp = dyn_cast<NamedSequenceOp>(parentOp)) {
+      return namedSequenceOp.getArgAttr(blockArg.getArgNumber(),
+                                        kArgConsumedAttrName) != nullptr;
+    }
+  }
+  return false;
+}
+
+struct TuningSpecsToMerge {
+  // - `namedSequenceOpsToMove`: Contains `NamedSequenceOp`s that either:
+  //     - Are not named `__kernel_config`.
+  //     - Do not have the `iree_codegen.tuning_spec_entrypoint` attribute.
+  SmallVector<NamedSequenceOp> namedSequenceOpsToMove;
+  // Maps a `NamedSequenceOp` to its user `ForeachMatchOp`, in which the
+  // `NamedSequenceOp` is used as a matcher or action inside a
+  // `ForeachMatchOp`.
+  llvm::DenseMap<NamedSequenceOp, ForeachMatchOp> namedSequenceToUser;
+};
+
+// Populates the mapping of `NamedSequenceOp` to `ForeachMatchOp`
+// by processing matchers and actions inside the given `ForeachMatchOp`.
+//
+// This function iterates over all matcher and action references in
+// `foreachMatch`. If a reference is a valid `SymbolRefAttr` that resolves to a
+// `NamedSequenceOp` within the `module`, the function records the association
+// by mapping the `NamedSequenceOp` to `foreachMatch` in `namedSequenceToUser`.
+static void updateForeachMatchMappings(
+    ForeachMatchOp foreachMatch, ModuleOp module,
+    llvm::DenseMap<NamedSequenceOp, ForeachMatchOp> &namedSequenceToUser) {
+  for (auto matcher : foreachMatch.getMatchers()) {
+    if (auto matcherSymRef = mlir::dyn_cast<mlir::SymbolRefAttr>(matcher)) {
+      if (auto matcherOp =
+              SymbolTable::lookupNearestSymbolFrom<NamedSequenceOp>(
+                  module, matcherSymRef)) {
+        namedSequenceToUser[matcherOp] = foreachMatch;
+      }
+    }
+  }
+
+  for (auto action : foreachMatch.getActions()) {
```

Comment:
Use the actual type here, since it's not obvious based on the context: https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable